libdnf/rpm/package_query.cpp

```diff
@@ -2453,6 +2456,56 @@ void PackageQuery::filter_recent(const time_t timestamp) {
     }
 }
 
+void PackageQuery::filter_unneeded() {
+    auto & pool = get_rpm_pool(p_impl->base);
+
+    // get a list of all installed packages
+    auto * installed_repo = pool->installed;
+    if (installed_repo == nullptr) {
+        (*p_impl).clear();
+        return;
+    }
+
+    libdnf::solv::SolvMap installed_solv_map(pool.get_nsolvables());
```

Comment:
What about to use a PackageQuery instead? It could replace `installed_solv_map` and `installed_pset`
```
libdnf::rpm::PackageQuery installed{p_impl->base, FLAG};
installed.filter_installed()
```